pkg/multicluster/kubeconfig_handler.go

```diff
@@ -0,0 +1,69 @@
+package multicluster
+
+import (
+	"context"
+	"sync"
+
+	"github.com/solo-io/go-utils/errors"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/factory"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/kube/cache"
+	"github.com/solo-io/solo-kit/pkg/multicluster/secretconverter"
+	v1 "github.com/solo-io/solo-kit/pkg/multicluster/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type ClusterId string
+
+type KubeConfigs map[ClusterId]*v1.KubeConfig
+type OnNewKubeConfigs func(updated KubeConfigs) error
+
+type KubeConfigHandler struct {
+	access   sync.Mutex
+	start    func(ctx context.Context) (<-chan error, error)
+	callback OnNewKubeConfigs
+}
+
+func NewKubeConfigHandler(kube kubernetes.Interface, cache cache.KubeCoreCache) (*KubeConfigHandler, error) {
+	kubeConfigClient, err := v1.NewKubeConfigClient(&factory.KubeSecretClientFactory{
+		Clientset:       kube,
+		Cache:           cache,
+		SecretConverter: &secretconverter.KubeConfigSecretConverter{},
+	})
+	if err != nil {
+		return nil, err
+	}
+	handler := &KubeConfigHandler{}
+	emitter := v1.NewKubeconfigsEmitter(kubeConfigClient)
+	eventLoop := v1.NewKubeconfigsEventLoop(emitter, &kubeConfigSyncer{handler: handler})
+	handler.start = func(ctx context.Context) (errors <-chan error, e error) {
+		return eventLoop.Run(nil, clients.WatchOpts{Ctx: ctx})
+	}
+	return handler, nil
+}
+
+func (h *KubeConfigHandler) SetCallback(callback OnNewKubeConfigs) {
+	h.access.Lock()
+	defer h.access.Unlock()
+	h.callback = callback
+}
+
+func (h *KubeConfigHandler) Start(ctx context.Context) (<-chan error, error) {
+	return h.start(ctx)
+}
+
+type kubeConfigSyncer struct {
+	handler *KubeConfigHandler
+}
+
+func (s *kubeConfigSyncer) Sync(_ context.Context, snap *v1.KubeconfigsSnapshot) error {
+	cfgs := snap.Kubeconfigs.List()
+	kubeConfigs := make(KubeConfigs)
+	for _, cfg := range cfgs {
+		kubeConfigs[ClusterId(cfg.GetMetadata().Name)] = cfg
+	}
+	if s.handler.callback == nil {
```

Comment:
do we need a lock? a new map is initialized every time

i can move the callback to constructor. i figured it would give the caller more flexibility, but i can move it to constructor if you think that's preferable

Reply:
if SetCallback uses a lock to save the callback, a lock must be used when accessing it.
so either add a GetCallback() that returns it under the lock. or,
set it in the ctor to avoid all the sync stuff. 


It seems that not having a callback is an error so i think setting in ctor makes sense.